superset-frontend/plugins/plugin-chart-echarts/src/Heatmap/transformProps.ts

```diff
@@ -0,0 +1,643 @@
+      );
+
+      yMaxes.set(
+        yCategoryValue,
+        Math.max(yMaxes.get(yCategoryValue) || Number.MIN_SAFE_INTEGER, value),
+      );
+
+      // TODO is there another way to do this? vv
+      // Collect x or y or all values when needed?
+      // Probably but then reiterating over entire dataset again.
+
+      xValues.set(
+        xCategoryValue,
+        (xValues.get(xCategoryValue) || []).concat(value),
+      );
+
+      yValues.set(
+        yCategoryValue,
+        (yValues.get(yCategoryValue) || []).concat(value),
+      );
+    });
+  }
+
+  // minBound = overallMinValue unless one of Min/Max bound is set
+  // (in which case it's the set minV bound, or 0 by default) or
+  // if normalized is checked, in which case 0 is lower default bound
+  let minBound = Number.isNaN(overallMinValue) ? 0 : overallMinValue;
+  let maxBound = Number.isNaN(overallMaxValue) ? 0 : overallMaxValue;
+
+  if (normalizeAcross === 'heatmap') {
+    if (yAxisBounds[0] != null) {
+      minBound = yAxisBounds[0];
+    } else if (normalized || yAxisBounds[1] != null) {
+      minBound = 0;
+    }
+
+    if (yAxisBounds[1] != null) {
+      maxBound = yAxisBounds[1];
+    } else if (normalized || yAxisBounds[0] != null) {
+      maxBound = 100;
+    }
+  }
+
+  const xCategories = sortCategoryList(sortXAxis, 'x', xSums);
+  const yCategories = sortCategoryList(sortYAxis, 'y', ySums);
+
+  // TODO this is awful!?
+  // key = x/ycategory or heatmap, (key = cell value, value = rank as percent)
+  const percentRanks: Map<string, Map<number, number>> = new Map();
+  if (normalized) {
+    if (normalizeAcross === 'heatmap') {
+      percentRanks.set(
+        'heatmap',
+        getPercentRanks(data.map(datum => datum[metricName] as number)),
+      );
+    } else if (normalizeAcross === 'x') {
+      xCategories.forEach(xCategory => {
+        // TODO the || stuff here is shady.
+        // Confirm what xCategories values (DataRecordValue) look like
+        percentRanks.set(
+          xCategory as string,
+          getPercentRanks(xValues.get(xCategory as string) || []),
+        );
+      });
+    } else if (normalizeAcross === 'y') {
+      yCategories.forEach(yCategory => {
+        percentRanks.set(
+          yCategory as string,
+          getPercentRanks(yValues.get(yCategory as string) || []),
+        );
+      });
+    }
+  }
+
+  const enrichedData = data.map(datum => {
+    const value: number = datum[metricName] as number; // TODO sketchy casts?
+    // TODO this code v is duplicated
+    const xCategoryValue: string = extractGroupbyLabel({
+      datum,
+      groupby: xCategory,
+      coltypeMapping,
+      timeFormatter: getTimeFormatter(dateFormat),
+      numberFormatter,
+    });
+    const yCategoryValue: string = extractGroupbyLabel({
+      datum,
+      groupby: yCategory,
+      coltypeMapping,
+      timeFormatter: getTimeFormatter(dateFormat),
+      numberFormatter,
+    });
```

Comment:
I think this is fine to be duplicated, unless you make a helper function it might not be worth trying to simplify this